### PR TITLE
fix model.fit() slow with sample_weight

### PR DESCRIPTION
**Pull request by nershman**
An issue was opened at the tensorflow repository: https://github.com/tensorflow/tensorflow/issues/48965
And the related pull request was denied because this needs to be submitted to keras: https://github.com/tensorflow/tensorflow/pull/54368
I'm re-submitting it here since it has been 3 weeks and the original author hasn't submitted a pull-request at the keras repo.

This is a small bugfix to improve performance when using `sample_weight` in `model.fit()`. Currently list comprehension is used to check that there are no None is sample_weight. This creates a very noticeable delay in the start of the fitting procedure when comparing a model with vs. without weights added.  The list comprehensions are changed to array operations to improve performance.

### CODE REVIEW

File: keras/engine/training_utils.py [hunk 1]
```diff
@@ -71,10 +71,10 @@ def handle_partial_sample_weights(outputs, sample_weights, sample_weight_modes,
     Tuple of sample weights, one sample weight for every output, and booleans
     describing the raw sample weights.
   """
-  any_sample_weight = sample_weights is not None and any(
```
Comment (fchollet) [inline keras/engine/training_utils.py:74]: The original logic was confusing, but the new logic does not do what's intended.

What's intended is something like

```python
if sample_weights is None:  # Most common case
  any_sample_weight = False
  partial_sample_weight = False
elif instance(sample_weights, (tuple, list, dict)):  # Multi-output case
  # Whether at least one output has sample weights
  any_sample_weight = any(w is not None for w in sample_weights)
  # Whether at least one output is missing sample weights
  partial_sample_weight = any(w is None for w in sample_weights)
```

We should go with the logic above.

Have you identified the unit tests is charge of testing this part of the code? They may need to be updated.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/engine/training_utils.py (hunk 1 above)
Comment (nershman) [reply keras/engine/training_utils.py:74]: There's some unit tests in keras/tests for temporal weights but I haven't found any other specific to testing sample weights. `test_fit_generator_with_sample_weight(self):` and `test_fit_generator_with_partial_sample_weight(self):` in keras/tests/temporal_sample_weights_correctness_test.py . Other than that I found tests for `Model.fit()` and `TensorLikeDataAdapter()` (which reference the relevant function) in the keras/engine folder but none of them seem to test partial weights. Is it possible that some new tests need to be written?

Regarding logic, would the logic behind this be correct? (All I did was change the `.all()` to `.any()`
`
any_sample_weight = sample_weights is not None and (
      sample_weights.numpy() != None).any()
  partial_sample_weight = any_sample_weight and (
      sample_weights.numpy() == None).any()
`

* any_sample_weight: return True when sample_weights contains None
* partial_sample_weight: return True when sample_weights contains both None and non-None.

I see that you omitted the check for non-None from the partial_sample_weight (`partial_sample_weight = any_sample_weight and` in the original code), was this intentional or a typo?  Or are you suggesting chaning how the logic is structured in the function?

Comment (fchollet) [reply keras/engine/training_utils.py:74]: > any_sample_weight: return True when sample_weights contains None

No, it's True if sample_weights contains any non-None value. So it there's at least a sample weights array for one output.
 
> partial_sample_weight: return True when sample_weights contains both None and non-None.

Yes, it's True when some outputs have sample weights arrays and others do not.

> I see that you omitted the check for non-None from the partial_sample_weight

It should be

```
partial_sample_weight = any_sample_weight and any(w is None for w in sample_weights)
```

Can you please add a test that check the behavior for a model with 3 outputs and only some of them receive sample_weight (as a dict or list)?

**Comment by nershman** [comment]
I compared the speeds of the suggested new approach and it's actually worse. Gist: https://gist.github.com/nershman/d826e78bd41e67f64a20129c341e8916

Plus, after looking into the code more I don't think the issue is the list comprehension. It's more likely related to TensorLikeDataAdaptor setting `check_all_flat=True` when it calls the function. I also discovered the bug disappears when output is more than one dimension. ( just running model.fit() [y_train,y_train] and [weights,weights]  to two dim output). 

I think it's best if I investigate this more and create an issue. If I find a solution (and get some tests made) I'll re-open this pull request.

**Comment by nershman** [comment]
In the function `_process_tensorlike`, the np.ndarray gets converted to a one-dimension tensor. When `check_partial_sample_weights` is called it then iterates over every observation in the tensor, whereas the expected behavior is to iterate over outputs checking that each set of weights is not None (rather than every single observation). I ran Cprofile and you can see the difference between calling mode.fit() with `sample_weights` vs. wrapping sample weights in a list (`[sample_weight`]`).
Untitled-1

I tried adding a check to `_process_tensorlike` to expand dimensions for dim-1 nparrays ( [e70be7f](https://github.com/keras-team/keras/pull/16177/commits/e70be7f4d0bfa632ec56506933238a765d45b038) ) but it causes tests not to pass for `data_adapter_test`. It seems to me that when pd.Series is input the same behavior is already implemented: it's converted to a numpy array and then dimension is expanded.)

File: keras/engine/data_adapter.py [hunk 2]
```diff
@@ -1033,6 +1033,8 @@ def _convert_single_tensor(x):
 
     if isinstance(x, np.ndarray):
       dtype = None
+      if x.ndim == 1:
+        x = np.expand_dims(x, axis=-1)
```
Comment (fchollet) [inline keras/engine/data_adapter.py:1037]: The rank expansion here is not correct, in many cases we want to be able to pass scalar data.

What could be some other solutions?

**Comment by gbaned** [comment]
@nershman Can you please check @fchollet's comments and keep us posted ? Thanks!

**Comment by nershman** [comment]
> @nershman Can you please check @fchollet's comments and keep us posted ? Thanks!

Thanks for the bump, I didn't get notified by the comment.

So the function also needs to be able to take as input scalars or multiple scalars for multiple outputs? (some simple examples being shape[1,2] for two output layers and one observation, and shape[2,1] for one output layer with two observations). I want to make sure I understand so I don't find another improper solution, thanks. 

Also, do you know a function which calls _convert_single_tensor with scalars as input? I'll try to find where else the function is called so I can better understand.

**Comment by mayer79** [comment]
Any progress on this issue? I would love to see it fixed.

**Comment by gbaned** [comment]
@nershman Any update on this PR? Please. Thank you!

**Comment by nershman** [comment]
> @nershman Any update on this PR? Please. Thank you!
I have been waiting for a reply to my clarification.

**Comment by gbaned** [comment]
Hi @fchollet Any update on this PR? Please. Thank you!

**Comment by nershman** [comment]
> Hi @fchollet Any update on this PR? Please. Thank you!

Hey I have finally found some time to work on this recently. Hopefully will suggest a different solution soon.

**Comment by nershman** [comment]
Why did this get closed?
